File: haskell/glue/src/Glue/Lib/Math.hs
module Glue.Lib.Math where

import Glue.Env qualified as E
import Glue.Eval (Eval)
import Glue.IR (Frame)
import Glue.Lib.Math.Arithmetic qualified as Arithmetic
import Glue.Lib.Math.Const qualified as Const
import Glue.Lib.Math.Logarithmic qualified as Logarithmic
import Glue.Lib.Math.Power qualified as Power
import Glue.Lib.Math.Trigonometric qualified as Trigonometric
import Glue.Lib.Math.Utility qualified as Utility

math :: Frame Eval
math =
    E.unionFramesList
        [ Arithmetic.arithmetic
        , Const.const
        , Logarithmic.logarithmic
        , Trigonometric.trigonometric
        , Utility.utility
        , Power.power
        ]

File: dart/glue/lib/src/lib/math.dart
import 'package:glue/src/module.dart';
import 'math/arithmetic/arithmetic.dart';
import 'math/const.dart';
import 'math/logarithmic/logarithmic.dart';
import 'math/power/power.dart';
import 'math/trigonometric/trigonometric.dart';
import 'math/utility/utility.dart';

/// Math module - complete mathematical function library
/// Mirrors Haskell Glue.Lib.Math exactly

/// The math module containing all mathematical functions from all submodules
/// Mirrors Haskell Glue.Lib.Math.math exactly
final ModuleInfo math = nativeModule('ffi.math', [
  // Include all arithmetic functions (+, -, *, /, %)
  ...arithmetic.definitions,

  // Include all constants (pi, e)
  ...const_.definitions,

  // Include all logarithmic functions (lg, ln, log)
  ...logarithmic.definitions,

  // Include all power functions (exp, pow, sqrt)
  ...power.definitions,

  // Include all trigonometric functions (sin, cos, tan, asin, acos, atan)
  ...trigonometric.definitions,

  // Include all utility functions (abs, ceil, floor, max, min, round, trunc)
  ...utility.definitions,
]);
